pulp/apis/sas_api.py

```diff
@@ -0,0 +1,623 @@
+                try:
+                    os.remove(file)
+                except FileNotFoundError:
+                    pass
+
+        def _write_sol(self, filename, vs):
+            """Writes a SAS solution file"""
+            values = [(v.name, v.value()) for v in vs if v.value() is not None]
+            with open(filename, "w") as f:
+                f.write('_VAR_,_VALUE_\n')
+                for name, value in values:
+                    f.write(f"{name},{value}\n")
+            return True
+
+        def _read_solution(self, lp, primal_out, dual_out):
+            status = SOLSTATUS_TO_STATUS[self._macro.get("SOLUTION_STATUS", "ERROR")]
+
+            if self.proc == "OPTLP":
+                # TODO: Check whether there is better implementation than zip().
+                values = dict(zip(primal_out['_VAR_'], primal_out['_VALUE_']))
+                rc = dict(zip(primal_out['_VAR_'], primal_out['_R_COST_']))
+                lp.assignVarsVals(values)
+                lp.assignVarsDj(rc)
+
+                prices = dict(zip(dual_out['_ROW_'], dual_out['_VALUE_']))
+                slacks = dict(zip(dual_out['_ROW_'], dual_out['_ACTIVITY_']))
+                lp.assignConsPi(prices)
+                lp.assignConsSlack(slacks, activity=True)
+            else:
+                # Convert primal out data set to variable dictionary
+                # Use pandas functions for efficiency
+                values = dict(zip(primal_out['_VAR_'], primal_out['_VALUE_']))
+                lp.assignVarsVals(values)
+            lp.assignStatus(status)
+            return status
+
+        # TODO: Improve it without reading the mps file.
+        def _get_max_upload_len(self, fileName):
+            maxLen = 0
+            with open(fileName, 'r') as f:
+                for line in f.readlines():
+                    maxLen = max(maxLen, max([len(word) for word in line.split(" ")]))
+            return maxLen + 1
+
+        def actualSolve(self, lp):
+            """Solve a well formulated lp problem"""
+            log.debug("Running SAS")
+
+            if (self.cas_options == {}):
+                raise PulpSolverError("""SASCAS: Provide cas_options with
+                    {port: , host: , authinfo: }
+                    or {port: , host: , username: , password: }.""")
+
+            if len(lp.sos1) or len(lp.sos2):
+                raise PulpSolverError("SASCAS: Currently SAS doesn't support SOS1 and SOS2.")
+
+
+            proc = self.proc = "OPTMILP" if self.mip else "OPTLP"
+            # Get Obj Sense
+            if lp.sense == constants.LpMaximize:
+                self.solverOptions["objsense"] = "max"
+            elif lp.sense == constants.LpMinimize:
+                self.solverOptions["objsense"] = "min"
+            else:
+                raise PulpSolverError("SASCAS : Objective sense should be min or max.")
+
+            status = None
+            with redirect_stdout(SASLogWriter(self.msg)) as self._log_writer:
```

Comment:
this part has many indentations and ifs. Can it be refactored so that it's easier to read? for example breaking it into other methods. re-using the code with different arguments, etc.